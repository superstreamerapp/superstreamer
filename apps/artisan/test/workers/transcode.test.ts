import "bun";
import { describe, expect, test } from "bun:test";
import {
  getMatches,
  mergeInput,
  mergeStream,
} from "../../src/workers/transcode";

const ALL_FFPROBE_RESULT = {
  video: {
    "http://mock.com/asset.mp4": {
      height: 720,
      framerate: 24,
    },
  },
  audio: {
    "http://mock.com/asset.mp4": {
      language: "eng",
      channels: 2,
    },
  },
};

const NO_FFPROBE_RESULT = {
  video: {
    "http://mock.com/asset.mp4": {},
  },
  audio: {
    "http://mock.com/asset.mp4": {},
  },
};

describe("merge input", () => {
  test("should add probe results to input", () => {
    const input = mergeInput(
      {
        type: "video",
        path: "http://mock.com/asset.mp4",
      },
      ALL_FFPROBE_RESULT,
    );
    expect(input).toMatchSnapshot();
  });

  test("should fail when no default can be decided", () => {
    const fn = () =>
      mergeInput(
        {
          type: "video",
          path: "http://mock.com/asset.mp4",
        },
        NO_FFPROBE_RESULT,
      );
    expect(fn).toThrowError();
  });

  test("should prioritize user defined defaults", () => {
    const input = mergeInput(
      {
        type: "video",
        path: "http://mock.com/asset.mp4",
        height: 1,
        framerate: 1,
      },
      ALL_FFPROBE_RESULT,
    );
    expect(input).toMatchSnapshot();
  });
});

describe("merge stream", () => {
  test("should not match when not of the same type", () => {
    const stream = mergeStream(
      {
        type: "video",
        codec: "h264",
        height: 720,
      },
      {
        type: "text",
        path: "http://mock.com/asset.mp4",
        language: "eng",
      },
    );
    expect(stream).toBeNull();
  });

  test("should match the same type", () => {
    const stream = mergeStream(
      {
        type: "video",
        codec: "h264",
        height: 1080,
      },
      {
        type: "video",
        path: "http://mock.com/asset.mp4",
        height: 1080,
        framerate: 25,
      },
    );
    expect(stream).toMatchSnapshot();
  });

  test("should prioritize user defined stream properties and default where available", () => {
    const stream = mergeStream(
      {
        type: "video",
        codec: "h264",
        height: 480,
      },
      {
        type: "video",
        path: "http://mock.com/asset.mp4",
        height: 1080,
        framerate: 25,
      },
    );
    expect(stream).toMatchSnapshot();
  });

  test("should default audio bitrate", () => {
    const stream = mergeStream(
      {
        type: "audio",
        codec: "aac",
      },
      {
        type: "audio",
        path: "http://mock.com/asset.mp4",
        language: "nld",
        channels: 2,
      },
    );
    expect(stream).toMatchSnapshot();
  });
});

describe("get list of matches", () => {
  test("should match", () => {
    const matches = getMatches(
      [
        {
          type: "video",
          codec: "hevc",
          height: 1080,
        },
        {
          type: "video",
          codec: "h264",
          height: 720,
        },
        {
          type: "audio",
          codec: "eac3",
          channels: 100,
          bitrate: 1_000_000,
        },
        {
          type: "audio",
          codec: "ac3",
          channels: 6,
        },
        {
          type: "audio",
          codec: "aac",
          channels: 2,
          language: "eng",
        },
        {
          type: "audio",
          codec: "aac",
          language: "nld",
        },
      ],
      [
        {
          type: "video",
          path: "http://mock.com/asset.mp4",
          height: 1080,
          framerate: 25,
        },
        {
          type: "audio",
          path: "http://mock.com/asset.mp4",
          language: "eng",
          channels: 6,
        },
        {
          type: "audio",
          path: "http://mock.com/asset/audio-nld.mp4",
          language: "nld",
          channels: 2,
        },
      ],
    );
    expect(matches).toMatchSnapshot();
  });
});
