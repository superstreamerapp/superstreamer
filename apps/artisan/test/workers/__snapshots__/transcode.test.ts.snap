// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`merge input should add probe results to input 1`] = `
{
  "framerate": 24,
  "height": 720,
  "path": "http://mock.com/asset.mp4",
  "type": "video",
}
`;

exports[`merge input should prioritize user defined defaults 1`] = `
{
  "framerate": 1,
  "height": 1,
  "path": "http://mock.com/asset.mp4",
  "type": "video",
}
`;

exports[`merge stream should match the same type 1`] = `
{
  "bitrate": 4000000,
  "codec": "h264",
  "framerate": 25,
  "height": 1080,
  "type": "video",
}
`;

exports[`merge stream should prioritize user defined stream properties and default where available 1`] = `
{
  "bitrate": 1000000,
  "codec": "h264",
  "framerate": 25,
  "height": 480,
  "type": "video",
}
`;

exports[`merge stream should default audio bitrate 1`] = `
{
  "bitrate": 128000,
  "channels": 2,
  "codec": "aac",
  "language": "nld",
  "type": "audio",
}
`;

exports[`get list of matches should match 1`] = `
[
  [
    "video",
    {
      "bitrate": 2000000,
      "codec": "hevc",
      "framerate": 25,
      "height": 1080,
      "type": "video",
    },
    {
      "framerate": 25,
      "height": 1080,
      "path": "http://mock.com/asset.mp4",
      "type": "video",
    },
  ],
  [
    "video",
    {
      "bitrate": 2000000,
      "codec": "h264",
      "framerate": 25,
      "height": 720,
      "type": "video",
    },
    {
      "framerate": 25,
      "height": 1080,
      "path": "http://mock.com/asset.mp4",
      "type": "video",
    },
  ],
  [
    "audio",
    {
      "bitrate": 384000,
      "channels": 6,
      "codec": "ac3",
      "language": "eng",
      "type": "audio",
    },
    {
      "channels": 6,
      "language": "eng",
      "path": "http://mock.com/asset.mp4",
      "type": "audio",
    },
  ],
  [
    "audio",
    {
      "bitrate": 128000,
      "channels": 2,
      "codec": "aac",
      "language": "eng",
      "type": "audio",
    },
    {
      "channels": 6,
      "language": "eng",
      "path": "http://mock.com/asset.mp4",
      "type": "audio",
    },
  ],
  [
    "audio",
    {
      "bitrate": 128000,
      "channels": 2,
      "codec": "aac",
      "language": "nld",
      "type": "audio",
    },
    {
      "channels": 2,
      "language": "nld",
      "path": "http://mock.com/asset/audio-nld.mp4",
      "type": "audio",
    },
  ],
]
`;
